src/Primitives/src/StringSegment.cs

```diff
@@ -99,13 +73,13 @@ public string Value
         {
             get
             {
-                if (!HasValue)
+                if (HasValue)
```

Comment:
Why not eliminate the multiple return statements here too? Is that optimization only valid for methods?

Reply:
cc  @jaredpar @AndyAyersMS

Reply:
For reference:
JIT produces for the code as is
```asm
G_M56189_IG01:
       55                   push     rbp
       4883EC10             sub      rsp, 16
       488D6C2410           lea      rbp, [rsp+10H]
       48897DF0             mov      gword ptr [rbp-10H], rdi
       488975F8             mov      qword ptr [rbp-08H], rsi

G_M56189_IG02:
       48837DF000           cmp      gword ptr [rbp-10H], 0
       7413                 je       SHORT G_M56189_IG03
       488B7DF0             mov      rdi, gword ptr [rbp-10H]
       8B75F8               mov      esi, dword ptr [rbp-08H]
       8B55FC               mov      edx, dword ptr [rbp-04H]
       393F                 cmp      dword ptr [rdi], edi
       E8A6FD9FFF           call     String:Substring(int,int):ref:this
       EB02                 jmp      SHORT G_M56189_IG04

G_M56189_IG03:
       33C0                 xor      rax, rax

G_M56189_IG04:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret  
```

and for the variant with the null-conditional:
```asm
G_M56187_IG01:
       55                   push     rbp
       4883EC10             sub      rsp, 16
       488D6C2410           lea      rbp, [rsp+10H]
       48897DF0             mov      gword ptr [rbp-10H], rdi
       488975F8             mov      qword ptr [rbp-08H], rsi

G_M56187_IG02:
       488B7DF0             mov      rdi, gword ptr [rbp-10H]
       4885FF               test     rdi, rdi
       7504                 jne      SHORT G_M56187_IG03
       33C0                 xor      rax, rax
       EB0B                 jmp      SHORT G_M56187_IG04

G_M56187_IG03:
       8B75F8               mov      esi, dword ptr [rbp-08H]
       8B55FC               mov      edx, dword ptr [rbp-04H]
       E8A6FD9FFF           call     String:Substring(int,int):ref:this

G_M56187_IG04:
       90                   nop      

G_M56187_IG05:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret     
```

Looks "better", but the perf-numbers shows that it is slower. Same tendencies on win and linux.
Maybe it is the branch-predictor? Or the `test` after the `mov`. I don't know...

Reply:
> presume you mostly are testing the case where Buffer is not null?

Correct. Thx for the explanation!

Reply:
You did't happen to paste in tier0 code above, did you? With newer jits the comments above the code should mention this.

Reply:
This is optimized code, taken from 2.1.403, so pre-tiering (because of pretty print-out).

Reply:
When I look at `get_Value` I get this code (with 3.0, but 2.1 is similar):
```asm
; Assembly listing for method StringSegment:get_Value():ref:this
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
; optimized code
; rsp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  6,  4.50)     ref  ->  rdi         this class-hnd
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T01] (  2,  2   )     ref  ->  rsi       class-hnd "impAppendStmt"
;  V03 tmp2         [V03,T02] (  2,  2   )     int  ->  rax         "impAppendStmt"
;
; Lcl frame size = 0

G_M24824_IG01:
       nop      

G_M24824_IG02:
       cmp      gword ptr [rdi+8], 0
       je       SHORT G_M24824_IG04
       mov      rsi, gword ptr [rdi+8]
       mov      eax, dword ptr [rdi+16]
       mov      edx, dword ptr [rdi+20]
       mov      rdi, rsi
       mov      esi, eax
       mov      rax, 0xD1FFAB1E
       cmp      dword ptr [rdi], edi

G_M24824_IG03:
       rex.jmp  rax   // tail call String.Substring(...)

G_M24824_IG04:
       xor      rax, rax

G_M24824_IG05:
       ret      
```
(similar layout change using the alternate version). 

My only real complaints here are:
* the jit doesn't handle flowing arguments into the tail call as well as it could. While we generally must evaluate args left to right, here we have some freedom as none of the evaluations can cause exceptions, so we would be better off evaluating the leftmost arg last. If we did that we could eliminate the register shuffles.
* we fail to realize the null check leading into the tail call is unnecessary

The assembly you're showing above is for a method that takes a parameter, almost like `Substring` but not quite. I am curious what led the jit to produce the code you've shown.

Can you post the source for it?

Reply:
> Can you post the source for it?

```c#
static void Main(string[] args)
{
    var stringSegment = new StringSegment1("abc");
    string value = Foo(stringSegment);

    if (value == null) Environment.Exit(1);
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static string Foo(StringSegment stringSegment)
{
    return stringSegment.Value;
}
```
dasm for `Foo` was shown.